File: library/src/main/java/me/shkschneider/skeleton/helper/SharedPreferencesHelper.java
package me.shkschneider.skeleton.helper;

import android.content.Context;
import android.content.SharedPreferences;
import android.support.annotation.NonNull;
import android.support.v4.content.SharedPreferencesCompat;

import java.util.Map;
import java.util.Set;

import me.shkschneider.skeleton.java.MapHelper;

public class SharedPreferencesHelper {

    protected SharedPreferencesHelper() {
        // Empty
    }

    private static SharedPreferences get() {
        return ApplicationHelper.context().getSharedPreferences(ApplicationHelper.name(), Context.MODE_PRIVATE);
    }

    public static boolean contains(@NonNull final String key) {
        return get().contains(key);
    }

    public static void remove(@NonNull final String key) {
        SharedPreferencesCompat.EditorCompat.getInstance().apply(get().edit().remove(key));
    }

    public static void clear() {
        SharedPreferencesCompat.EditorCompat.getInstance().apply(get().edit().clear());
    }

    public static void dump() {
        final Map<String, ?> map = get().getAll();
        for (final String key : MapHelper.keys(map)) {
            LogHelper.debug(key + "=" + map.get(key).toString());
        }
    }

    // String

    public static boolean put(@NonNull final String key, final String value) {
        LogHelper.verbose(key + " = " + value);
        SharedPreferencesCompat.EditorCompat.getInstance().apply(get().edit().putString(key, value));
        return true;
    }

    public static String get(@NonNull final String key, final String defaultValue) {
        return get().getString(key, defaultValue);
    }

    // Set<String>

    public static boolean put(@NonNull final String key, final Set<String> value) {
        LogHelper.verbose(key + " = " + value);
        SharedPreferencesCompat.EditorCompat.getInstance().apply(get().edit().putStringSet(key, value));
        return true;
    }

    public static Set<String> get(@NonNull final String key, final Set<String> defaultValue) {
        return get().getStringSet(key, defaultValue);
    }

    // Int

    public static boolean put(@NonNull final String key, final Integer value) {
        LogHelper.verbose(key + " = " + value);
        SharedPreferencesCompat.EditorCompat.getInstance().apply(get().edit().putInt(key, value));
        return true;
    }

    public static Integer get(@NonNull final String key, final Integer defaultValue) {
        return get().getInt(key, defaultValue);
    }

    // Long

    public static boolean put(@NonNull final String key, final Long value) {
        LogHelper.verbose(key + " = " + value);
        SharedPreferencesCompat.EditorCompat.getInstance().apply(get().edit().putLong(key, value));
        return true;
    }

    public static Long get(@NonNull final String key, final Long defaultValue) {
        return get().getLong(key, defaultValue);
    }

    // Float

    public static boolean put(@NonNull final String key, final Float value) {
        LogHelper.verbose(key + " = " + value);
        SharedPreferencesCompat.EditorCompat.getInstance().apply(get().edit().putFloat(key, value));
        return true;
    }

    public static Float get(@NonNull final String key, final Float defaultValue) {
        return get().getFloat(key, defaultValue);
    }

    // Boolean

    public static boolean put(@NonNull final String key, final Boolean value) {
        LogHelper.verbose(key + " = " + value);
        SharedPreferencesCompat.EditorCompat.getInstance().apply(get().edit().putBoolean(key, value));
        return true;
    }

    public static Boolean get(@NonNull final String key, final Boolean defaultValue) {
        return get().getBoolean(key, defaultValue);
    }

}

File: android/src/main/kotlin/me/shkschneider/skeleton/android/core/data/SharedPreferencesHelper.kt
package me.shkschneider.skeleton.data

import android.content.Context
import android.content.SharedPreferences
import androidx.core.content.edit

import me.shkschneider.skeleton.helper.ApplicationHelper
import me.shkschneider.skeleton.helper.ContextHelper
import me.shkschneider.skeleton.helperx.log.Logger

object SharedPreferencesHelper {

    private fun editor(name: String?): SharedPreferences {
        return ContextHelper.applicationContext().getSharedPreferences(
                if (name.isNullOrBlank()) ApplicationHelper.packageName else name,
                Context.MODE_PRIVATE)
    }

    fun contains(key: String): Boolean {
        return editor(null).contains(key)
    }

    fun contains(key: String, name: String? = null): Boolean {
        return editor(name).contains(key)
    }

    fun remove(key: String, name: String? = null) {
        editor(name).edit {
            remove(key)
        }
    }

    fun clear(name: String? = null) {
        editor(name).edit {
            clear()
        }
    }

    // String

    fun putString(key: String, value: String, name: String? = null) {
        Logger.verbose("$name: $key = $value")
        editor(name).edit {
            putString(key, value)
        }
    }

    fun getString(key: String, defaultValue: String, name: String? = null): String? {
        return editor(name).getString(key, defaultValue)
    }

    // Set<String>

    fun putStringSet(key: String, value: Set<String>, name: String? = null) {
        Logger.verbose("$name: $key = $value")
        editor(name).edit {
            putStringSet(key, value)
        }
    }

    fun getStringSet(key: String, defaultValue: Set<String>, name: String? = null): Set<String>? {
        return editor(name).getStringSet(key, defaultValue)
    }

    // Integer

    fun putInteger(key: String, value: Int, name: String? = null) {
        Logger.verbose("$name: $key = $value")
        editor(name).edit {
            putInt(key, value)
        }
    }

    fun getInteger(key: String, defaultValue: Int, name: String? = null): Int? {
        return editor(name).getInt(key, defaultValue)
    }

    // Long

    fun putLong(key: String, value: Long, name: String? = null) {
        Logger.verbose("$name: $key = $value")
        editor(name).edit {
            putLong(key, value)
        }
    }

    fun getLong(key: String, defaultValue: Long, name: String? = null): Long? {
        return editor(name).getLong(key, defaultValue)
    }

    // Float

    fun putFloat(key: String, value: Float, name: String? = null) {
        Logger.verbose("$name: $key = $value")
        editor(name).edit {
            putFloat(key, value)
        }
    }

    fun getFloat(key: String, defaultValue: Float, name: String? = null): Float? {
        return editor(name).getFloat(key, defaultValue)
    }

    // Boolean

    fun putBoolean(key: String, value: Boolean, name: String? = null) {
        Logger.verbose("$name: $key = $value")
        editor(name).edit {
            putBoolean(key, value)
        }
    }

    fun getBoolean(key: String, defaultValue: Boolean, name: String? = null): Boolean? {
        return editor(name).getBoolean(key, defaultValue)
    }

}
